pkg/controller/pxc/backup.go

```diff
@@ -73,73 +74,58 @@ func (r *ReconcilePerconaXtraDBCluster) reconcileBackups(cr *api.PerconaXtraDBCl
 				return fmt.Errorf("storage %s doesn't exist", bcp.StorageName)
 			}
 
-			bcpjob, err := bcpObj.Scheduled(&bcp, strg, operatorPod)
-			if err != nil {
-				return fmt.Errorf("unable to schedule backup: %w", err)
-			}
-			err = setControllerReference(cr, bcpjob, r.scheme)
-			if err != nil {
-				return fmt.Errorf("set owner ref to backup %s: %v", bcp.Name, err)
-			}
-
-			// Check if this Job already exists
-			currentBcpJob := new(batchv1beta1.CronJob)
-			err = r.client.Get(context.TODO(), types.NamespacedName{
-				Name:      bcpjob.Name,
-				Namespace: bcpjob.Namespace,
-			}, currentBcpJob)
-			if err != nil && !k8serrors.IsNotFound(err) {
-				return errors.Wrapf(err, "create scheduled backup %s", bcp.Name)
+			sch := BackupScheduleJob{}
+			schRaw, ok := r.crons.backupJobs.Load(bcp.Name)
+			if ok {
+				sch = schRaw.(BackupScheduleJob)
 			}
 
-			if k8serrors.IsNotFound(err) {
-				err = r.client.Create(context.TODO(), bcpjob)
+			if !ok || sch.PXCScheduledBackupSchedule.Schedule != bcp.Schedule ||
+				sch.PXCScheduledBackupSchedule.StorageName != bcp.StorageName {
+				r.log.Info("Creating or updating backup job", "name", bcp.Name, "schedule", bcp.Schedule)
+				r.deleteBackupJob(bcp.Name)
+				jobID, err := r.crons.crons.AddFunc(bcp.Schedule, r.createBackupJob(cr, bcp, strg.Type))
 				if err != nil {
-					return errors.Wrapf(err, "create scheduled backup %s", bcp.Name)
-				}
-			} else if !reflect.DeepEqual(currentBcpJob.Spec, bcpjob.Spec) {
-				err = r.client.Update(context.TODO(), bcpjob)
-				if err != nil {
-					return errors.Wrapf(err, "update backup schedule %s", bcp.Name)
+					return errors.Wrap(err, "failed to add backup cron job")
 				}
+
+				r.crons.backupJobs.Store(bcp.Name, BackupScheduleJob{
+					PXCScheduledBackupSchedule: bcp,
+					JobID:                      jobID,
+				})
 			}
 		}
 	}
 
-	// Reconcile backups list
-	bcpList := batchv1beta1.CronJobList{}
-	err = r.client.List(context.TODO(),
-		&bcpList,
-		&client.ListOptions{
-			Namespace: operatorPod.ObjectMeta.Namespace,
-			LabelSelector: labels.SelectorFromSet(map[string]string{
-				"cluster": cr.Name,
-				"type":    "cron",
-			}),
-		},
-	)
-	if err != nil {
-		return fmt.Errorf("get backups list: %v", err)
-	}
+	var backupDeleteError error
 
-	for _, item := range bcpList.Items {
+	r.crons.backupJobs.Range(func(k, v interface{}) bool {
+		item := v.(BackupScheduleJob)
+		if !strings.HasPrefix(item.Name, backupNamePrefix) {
+			return true
+		}
 		if spec, ok := backups[item.Name]; ok {
 			if spec.Keep > 0 {
 				oldjobs, err := r.oldScheduledBackups(cr, item.Name, spec.Keep)
 				if err != nil {
-					return fmt.Errorf("remove old backups: %v", err)
+					backupDeleteError = errors.Wrap(err, "remove old backup")
+					return false
 				}
 
 				for _, todel := range oldjobs {
 					_ = r.client.Delete(context.TODO(), &todel)
```

Comment:
Shouldn't the error here be handled?

Reply:
because we can't exit right here, i assign error to separate variable, break the loop and return this error at the end of the function

Reply:
Oh, that comments layout... Sorry, it was about line 134, not 114-115. It's not your code, but nevertheless, r.client.Delete() can return error

Reply:
7ea73a9b